text/0000-placement-traits.md

```diff
@@ -0,0 +1,362 @@
+    /// `Owner` is the type of the end value of both `PLACE <- EXPR` and
+    /// `box EXPR`.
+    ///
+    /// Note that when `PLACE <- EXPR` is solely used for side-effecting an
+    /// existing data-structure, e.g. `Vec::back`, then `Owner` need not carry
+    /// any information at all (e.g. it can be the unit type `()` in that
+    /// case).
+    type Owner;
+
+    /// `Data` is the type of the value to be emplaced.
+    type Data: ?Sized;
+
+    /// Returns the address where the input value will be written.
+    /// Note that the data at this address is generally uninitialized,
+    /// and thus one should use `ptr::write` for initializing it.
+    fn pointer(&mut self) -> *mut Self::Data;
+
+    /// Converts self into the final value, shifting deallocation/cleanup
+    /// responsibilities (if any remain), over to the returned instance of
+    /// `Owner` and forgetting self.
+    unsafe fn finalize(self) -> Self::Owner;
+}
+
+/// All types implemeinting this trait can be constructed using using the `box`
+/// keyword (iff the associated `Allocator` implements `Default`) or by placing
+/// (`<-`) a value into an instance of the associated `Allocator`.
+///
+/// ```norun
+/// let boxed1: Box<i32, SomeAllocator> = box 1; // where SomeAllocator: Default
+/// let boxed2: Box<i32, SomeAllocator> = SomeAllocator::new() <- 1;
+/// ```
+pub trait Allocate<Data: ?Sized>: Sized {
+    /// The place that will negotiate the storage of the data.
+    type Place: Place<Data=Data, Owner=Self>;
+    /// The allocator for this type.
+    type Allocator: Allocator;
+
+    /// Create a `Place` for `Self` with the given `Allocator`.
+    fn allocate(with: Self::Allocator) -> Self::Place
+        where Data: Sized;
+}
+
+```
+
+Additionally, we define the following blanket impls to allow anything
+implementing `Allocate` to be allocated in any `Allocator`. When specialization
+lands, these can be marked default.
+
+```rust
+// Allows `let owner: T = ALLOCATOR <- thing;`
+impl<T, D, A> Placer<D, A> for T
+    where T: Allocator,
+          A: Allocate<D, Allocator=T>
+{
+    type Place = A::Place;
+
+    fn make_place(self) -> Self::Place
+        where Self: Sized
+    {
+        A::allocate(self)
+    }
+}
+
+// Allows `let owner: T = box thing;`
+impl<A, D> Boxer<D> for A
+    where A: Allocate<D>,
+          A::Allocator: Default
+{
+    type Place = A::Place;
+
+    fn make_place() -> Self::Place
+        where Self: Sized
+    {
+        A::allocate(Default::default())
+    }
+}
+```
+
+Finally, to support DST placement, this RFC explicitly loosens the placement
+protocol guarantees. Specifically, the place in placement in/new is not
+guaranteed to be allocated before the evaluation of the expression on the right
+hand side. DST placement needs this to be able to compute the size of the DST
+before allocating the place. This means that, in the following cases, whether or
+not the `Box` is ever allocated is explicitly undefined:
```

Comment:
hmm that seems like a subtle distinction you are making versus what the text says ... that is, I read the phrase "the place in placement in/new" as talking about the LHS of `P <- V` ... is there some way to rephrase to make this clearer?